Add new quest to ask if a place is suitable for vegetarians or vegans

ENT8R opened the pull request:
Here is the updated version of #499. The quest is now split into two seperate quests and each quest is a simple yes-no-dialog with an "Other answer" for "Only serves vegetarian/vegan food"
The vegetarian question looks like this:
<img src="https://user-images.githubusercontent.com/25306497/29487515-bc73472a-84fa-11e7-95aa-68f8901b8b7d.png" width="325px"> <img src="https://user-images.githubusercontent.com/25306497/29487517-bfb6e7ca-84fa-11e7-85f2-ed08ecc90462.png" width="325px">

And the vegan question looks like this:

<img src="https://user-images.githubusercontent.com/25306497/29487521-c34708ca-84fa-11e7-9b5c-1ec3b89b0495.png" width="325px"> <img src="https://user-images.githubusercontent.com/25306497/29487523-c669365e-84fa-11e7-8309-95f737f7c976.png" width="325px">

westnordost commented:
"suitable" invites somewhat subjective classification. Is this not defined any better?

What is the rationale behind putting that answer into "Other answers"?

Do you have any suggestion for icon motives?

ENT8R commented:
> Do you have any suggestion for icon motives?

I would propose either the [fork and knife emoji from twemoji](https://commons.wikimedia.org/wiki/File:Twemoji2_1f374.svg) or the [fork, knife and spoon emoji from twemoji](https://commons.wikimedia.org/wiki/File:Twemoji2_1f37d.svg), because I think, there is nothing that could describe vegan or vegetarian...

> "suitable" invites somewhat subjective classification. Is this not defined any better?

I have no idea, which other word we can use for that... Maybe an explanatory text like "Does this place explicitly offer no fish or meat?" for the vegetarian quest and "Does this place explicitly offer no fish, meat, dairy, eggs or any other animal product?" for the vegan quest. What do you think?

> What is the rationale behind putting that answer into "Other answers"?

Because it would take too much place in the buttonpanel...

westnordost commented:
Well I was thinking a leaf for either of the two and something else for the other one (or both).

> I have no idea, which other word we can use for that... 

Nothing helpful in the wiki? Your proposed questions are too long. Id propose "Does ... explicitly offer vegetarian food?" (instead of "suitable") but I don't know if this fits the definition in the wiki.

> Because it would take too much place in the buttonpanel...
Even just "Only" / "Ausschließlich"?

ENT8R commented:
> Even just "Only" / "Ausschließlich"?

Maybe this could work, but the german word is much more longer than the english one... This can be a problem, because it could be displayed as:

AUS
SCHLIEß
LICH

especially on small devices... But I will test it anyway...

> Nothing helpful in the wiki?

No, not really

ENT8R commented:
> Id propose "Does ... explicitly offer vegetarian food?" (instead of "suitable") but I don't know if this fits the definition in the wiki.

Yes, this fits the definition in the wiki. I am implementing this right now.

matkoniecz commented:
I would replace

> nodes, ways with ( amenity ~ restaurant|cafe|pub|fast_food|bar ) and diet:vegetarian and !diet:vegan

by

> nodes, ways with ( amenity ~ restaurant|cafe|pub|fast_food|bar ) and diet:vegetarian=yes and !diet:vegan

diet:vegetarian=no imply diet:vegan=no

ENT8R commented:
> diet:vegetarian=no imply diet:vegan=no

Should we add `diet:vegan=no` when, in vegetarian quest, No is clicked?

matkoniecz commented:
> Should we add diet:vegan=no when, in vegetarian quest, No is clicked?

Probably not, otherwise one should add also `diet:fruitarian=no`, `diet:ovo_vegetarian=no`, `diet:lacto_vegetarian=no`

>> Nothing helpful in the wiki?
> No, not really

wiki has 

>     full options available : provides at least a proper choice for this option, constantly available 
    (i.e full meal in a restaurant, specific section in shop)

https://wiki.openstreetmap.org/wiki/Key:diet

"constantly available" requirement may make tagging this tricky.

ENT8R commented:
> Even just "Only" / "Ausschließlich"?

I tested it right now, and contrary to my expections, it is possible to implement a button with "Only" in english and even "Ausschließlich" in german. It is not displayed like I said in the comment above.

matkoniecz commented:
ops, "only" - I updated https://github.com/westnordost/StreetComplete/pull/506#issuecomment-323544900

ENT8R commented:
> Nothing helpful in the wiki?

I changed the question to "Does [..] explicitly offer vegan/vegetarian food?" I think this is much better!

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/quests/diet_type/AddVegan.java:
```diff
@@ -0,0 +1,43 @@
+
+	@Override protected String getTagFilters()
+	{
+		return "nodes, ways with ( amenity ~ restaurant|cafe|pub|fast_food|bar ) and diet:vegetarian and !diet:vegan";
```
`diet:vegetarian~yes|only`
Because this quest shouldn't show if the user selected "no" in the vegetarian quest

Also, superfluous parenthesis.

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/quests/diet_type/AddVeganForm.java:
```diff
@@ -0,0 +1,51 @@
+		}
+		else
+		{
+			setTitle(R.string.quest_dietType_vegan_title);
```
As usual, always require a name please. (So, add ` and name` in the query)

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/quests/diet_type/AddVegetarian.java:
```diff
@@ -0,0 +1,43 @@
+
+	@Override protected String getTagFilters()
+	{
+		return "nodes, ways with ( amenity ~ restaurant|cafe|pub|fast_food|bar ) and !diet:vegetarian";
```
I need to ask here: Why bars and pubs? These are not primarily eating places. Especially bars not. (False friends for Germans. `amenity=bar` is not a German "Bar", that's a `amenity=pub`). `amenity=bar` is something like a coctail bar.

ENT8R commented:
Everything you proposed is now implemented:
- Other question: ("Does [..] explicitly offer vegan/vegetarian food?")
- Reordered answers: Only/Ausschließlich is now an extra button next to Yes and No
- Removed `pub` and `bar` from the query
- Added `name` to the query
- Removed superfluous parenthesis

westnordost commented:
Okay, cool. Aren't the two forms now not exactly the same though? You can set the title from the outside (in `createForm()` method).

I can only merge this after I created an icon for this and implemented some country whitelist/blacklist. Remember, I talked about that this quest should only be activated in a certain set of countries (first). I.e. in detail, it should only be activated 
- In "meat-countries": Countries where it is common that most/many dishes will have meat in them. So, these are usually countries with both an industrialized agriculture (so probably not most of Africa) and where eating meat is seen as a desirable thing (so not India and Bhuddist countries I'd say)
- (And thus where) veganism / vegetarianism is culturally a thing
Question is whether it should rather be a white- or blacklist. I usually prefer a whitelist.

ENT8R commented:
> Aren't the two forms now not exactly the same though? You can set the title from the outside (in createForm() method)

This is implemented now!

rugk commented:
> Well I was thinking a leaf for either of the two and something else for the other one (or both).

No, please don't. This clashes with https://github.com/westnordost/StreetComplete/issues/366 and/or https://github.com/westnordost/StreetComplete/issues/83. Better really use  :spoon:/:fork_and_knife: or so… I doubt you'll creating another quest related to eatery… :wink:

rugk commented:
As for an explanatory text to clarify when one should select yes (because that is not obvious when a restaurant e.g. offers one vegetarian dish and the rest is just with meat) I'd suggest to use the text, which is used in @Rudloff's [OpenVegeMap](https://openvegemap.netlib.re/):
> If in doubt: only places that provide a proper choice for an option should be tagged _yes_. For instance, a restaurant with vegetarian starters but no vegetarian main course is to be tagged _no_. Same for, a pub that can't serve more than light snacks (baked potato, green salad) to vegetarians. 

(shorten it a bit and use it as a small grey explanatory text)

ENT8R commented:
Good idea! I will implement this now!

rugk commented:
The icon? Okay, but use a proper [fork & knife](https://openclipart.org/search/?query=fork+knife) icon as a template and then just apply a color filter or so…

ENT8R commented:
> The icon?

No no, I mean the explanatory text.

ENT8R commented:
>> Aren't the two forms now not exactly the same though? You can set the title from the outside (in `createForm()` method)

> This is implemented now!

I think I have to revert the changes I made, because I can't get the `name` tag of the building from outside the form.
At the moment I'm trying to do this:
`public AbstractQuestAnswerFragment createForm()
      {
		AbstractQuestAnswerFragment form =  new AddDietTypeForm();
		String name = form.getElementName();
		if(name != null)
		{
			form.setTitle(R.string.quest_dietType_vegan_name_title, name);
		}
		return form;
	}`
But if I execute this, the app crashes...
Do you have an idea how to do this?

westnordost commented:
Yes, that doesn't work that way. But doesn't matter, the way how the quest titles are set changes after when I merged the `undo `branch, then it is possible this way. So, just leave it that way, I will fix it after the merge.

However, such an explanatory text as rugk mentioned would be nice.

ENT8R commented:
Oh, this is perfect! Then I will push the latest changes now.

rugk commented:
BTW does it even make sense to ask for vegetarian for cafés? I mean I have never seen a cafe, where you can eat meat cake or bones. (possibly only galantine, but whcih cafe offers gummy bears?? And even if they did, they would still have enough dishes for vegetarians.)
Asking for vegan is of course fine, because of milk or so, but vegetarian? As I [mentioned before](https://github.com/westnordost/StreetComplete/issues/479#issuecomment-323132503) Starbucks is sometimes tagged as vegan, so uhmm… (Do they even sell anything else than :coffee: coffee?)

Rudloff commented:
Starbucks is probably tagged as vegan because you can ask for soy milk instead of cow milk in all their beverages (chai latte, capucino, etc.) and they also have some vegan cakes.
Most cafes are indeed vegetarian friendly but some of them serve breakfasts with things like bacon, sausage, etc.

westnordost commented:
> BTW does it even make sense to ask for vegetarian for cafés?

Yes it does.

rugk commented:
> Most cafes are indeed vegetarian friendly but some of them serve breakfasts with things like bacon, sausage, etc.

I would still tag them as vegetarian, anyway. Because you likely can also get that without sausages and so one, and you have enough vegetarian things to choose from. As I said, "there is a proper choice of dishes", or as you explain on your site, the main dish for a cafe is coffee and cake and similar things. These are all vegetarian – it does not really mater that you can also get some bacon for breakfast there then.

Rudloff commented:
IMHO this is a discussion that needs to happen on the OSM wiki so the tags can be more precisely defined.

rugk commented:
@Rudloff Sure, we can also discuss something else, but I see that as a StreetComplete problem. The tags are all right. But in StreetComplete always a low number of false-positives is desirable, i.e. quests should not be shown in cases where it makes no sense, e.g. because they are always true (as the meaning is implied). That's the whole reason why I am asking. If we would exclude cafès from the vegetarian quest, (but include it in the vegan quest) then we would not have to ask that (IMHO, useless) question to surveyors.
I understand the tags as "suitable for vegetarians" and also base that assumption on your description on your website. I think this is okay/reasonable, as there is also an "only" tag, so really… the OSM part is okay.
And according to your description this also applies to Starbucks (milk and vegan cakes). In reality **you can not get a cafe, where you cannot get a proper choice of vegetarian products**. They are always vegetarian.

Rudloff commented:
Well in that case, even I do think it is useful to tag cafes, I understand your point and won't oppose not asking if they are vegetarian in StreetComplete.

westnordost commented:
Correction to what I said in the commit comment: It would work, since it is nominative, but I'd like to not add strings where they are not necessary.

westnordost commented:
So, this PR is blocked by defining a country whitelist of where the quest should be shown. See https://github.com/westnordost/StreetComplete/pull/506#issuecomment-323579888

Rudloff commented:
Wikipedia has [some data](https://en.wikipedia.org/wiki/Vegetarianism_by_country) on the percentage of vegetarians per country. Maybe it can help (if a country has data it means that vegetarianism is a thing there?).

rugk commented:
Great idea.

rugk commented:
I created the YAML file based on the Wikipedia data there: https://gist.github.com/rugk/aac00507c825ccdb5df0526b09007942

However, this of course only includes ethical vegetarianism. In some countries people may just not afford meat – at least in restaurants. Although I am not sure there are much restaurants (except for tourism) or so… However, for tourism the local data (I gathered from Wikipedia here) would not matter either, again…

rugk commented on app/src/main/res/values/strings.xml:
```diff
@@ -388,4 +388,10 @@ Otherwise, you can download another keyboard in the app store. Popular keyboards
+    <string name="quest_dietType_vegetarian_name_title">Does "%s" explicitly offer vegetarian food?</string>
+    <string name="quest_dietType_vegan_name_title">Does "%s" explicitly offer vegan food?</string>
+    <string name="quest_hasFeature_only">Only</string>
+    <string name="quest_dietType_explanation_vegan">Vegan meals contain no animal products (no meat, no milk products, no eggs,…).</string>
```
A space before  `…`/after the comma is missing.

rugk commented on app/src/main/res/values/strings.xml:
```diff
@@ -388,4 +388,10 @@ Otherwise, you can download another keyboard in the app store. Popular keyboards
     <string name="quest_parkingType_surface">Normal</string>
     <string name="quest_parkingType_underground">Underground garage</string>
     <string name="quest_parkingType_multiStorage">Car park</string>
+    <string name="quest_dietType_vegetarian_name_title">Does "%s" explicitly offer vegetarian food?</string>
```
I have a problem with "explicitly" here. `quest_dietType_explanation` may hopefully be sufficient, but with this "explicitly" I would not tag a restaurant offering 2-4 meals as "vegetarian". Is that intended? Would not 2-4 meals be enough for a restaurant?

Rudloff replied:
I understood "explicitly" as meaning that vegetarian meals are on the menu and you don't need to ask for them (e.g. "Can I have this but without the meat, please?").

rugk replied:
Ah, that's a different way I understood it. Maybe change it to "offer on it's menu"?

rugk commented:
As for the icon, you do not want to use the colored :fork_and_knife: I proposed earlier?

I see you created [this icon](https://wiki.openstreetmap.org/wiki/File:StreetComplete_quest_restaurantveg.svg) obviously for this quest, but still we need two icons: One for vegan and one for vegetarian. And if we just add a "+" to vegetarian and "++" to vegan, or the same thing with starts or maybe even with carrots or so, but there should be a visual difference.

rugk commented:
BTW this PR is now unblocked.

habi commented:
A logo for the vegetarian places could be http://vegetarianmark.net

rory commented:
Perhaps I'm too late to the conversation, but I have some feedback on the (English) wording.

I think the original wording of "suitable for vegetarians" is the correct English term for what `diet:vegetarian=yes` means. It means there is at least one proper vegetarian option available. Asking "Does "%s" explicitly offer vegetarian food?" seems weird and, honestly, a little aggressive. "Does %s offer vegetarian food?" is better, but not as good as "suitable for vegetarians". Likewise for vegan.

I suggest some form of V for a logo for vegetarian, since that's commonly used on menus or on food in supermarkets in my experience. But I'm not sure what do for vegan.

ENT8R commented:
I will change the wording a little bit.

westnordost commented:
Hmm wait, that the wording was set to "explicitly" was result of the prior discussion. @rory did neither refute nor even respond these arguments.

To recap, the reason why "explicitly" is mentioned (and explicitly can mean as little as that on their menu, they have a section "Vegetarian menus"), is, that if it is not explicitly mentioned that something is vegetarian, one cannot assume that a particular dish is vegetarian.

I.e. for "Bratkartoffeln" (fried potatoes) in Germany, it can be the case, that there is some ham inside as well, without that it is explicitly mentioned on the menu. Many dishes that seem to be vegetarian / vegan may not actually be. It should not be in the responsibility of the surveyor to sort this out (by interviewing the chef??).

I would find it much worse to tag a place wrongly as `diet:vegetarian=yes` as to place a tag wrongly as `diet:vegetarian=no`.

Of course, thinking of India, where vegetarian restaurants are more the norm than not, no restaurant will actually "explicitly" mention that they have veg food (because everything on their menu is). But currently, I think the users will have enough common sense to tag a "only-vegetarian-without-explicitly-mentioning-it"-restaurant not as `diet:vegetarian=no`.

Just removing the "explicitly" is not the right solution. I was thinking of "...caters to..." but I am not native speaker, so it could be that "caters to" is actually even harder than "explicitly".

rory commented:
You're right, I didn't address the original points. However I would content that "Is [place] suitable for vegetarians?" is exactly what `diet:vegetarian=yes` means. That there is at least a full vegetarian meal option available. Maybe it's just clear to me as a native English speaker....

ENT8R commented:
```
Hmm wait, that the wording was set to "explicitly" was result of the prior discussion.
```
:grin::joy: It is so long ago that I created this PR that I forgot that...
